### PR TITLE
Room DB Integration for Product List

### PR DESCRIPTION
**Pull request by SangeethaTestPress**
### Changes done
- Added kotlin plugins to store
- Added Room DB to store products list
- Added MultiDex implementation
- Removed ProductsItem data class since it is already present in the Product Detail Response

### Reason for the changes
-  the total number of references that can be invoked by the code within a single Dalvik Executable (DEX) bytecode file exceed  65536 [https://developer.android.com/studio/build/multidex]
- Added Room to support offline access.

### CODE REVIEW

File: core/src/main/java/in/testpress/database/ProductsListDao.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+@Dao
+interface ProductsListDao: BaseDao<ProductsListEntity> {
+    @Transaction
+    @Query("SELECT * FROM productslistentity")
+    fun getAll(): LiveData<ProductsListEntity>
+
+    @Query("SELECT * from productslistentity where id = :id LIMIT 1")
+    fun findById(id: Long): LiveData<ProductsListEntity>
+}
```
Comment (Karthik-0) [inline core/src/main/java/in/testpress/database/ProductsListDao.kt:15]: please leave one line at the end of file

File: core/src/androidTest/java/in/testpress/core/database/ProductsListDaoTest.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Test
+    fun getProductsList() {
+        val productsList = createProductList()
+        db.productsListDao().insert(productsList)
+        val fetchedProducts = db.productsListDao().getAll().getOrAwaitValue()
+        Assert.assertEquals(productsList,fetchedProducts)
+    }
+
+}
```
Comment (Karthik-0) [inline core/src/androidTest/java/in/testpress/core/database/ProductsListDaoTest.kt:56]: please leave one line at the end of file

File: core/src/main/java/in/testpress/util/Converters.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        return Gson().fromJson(value, listType)
+    }
+
+    @TypeConverter
+    @JvmStatic
+    fun productsItemToString(products: List<ProductsItem?>?): String? {
+        val gson = Gson()
+        return gson.toJson(products)
+    }
+}
```
Comment (Karthik-0) [inline core/src/main/java/in/testpress/util/Converters.kt:38]: please leave one line at the end of file

**Review by Karthik-0 — APPROVED** [review]

File: core/src/androidTest/java/in/testpress/core/database/ProductsListDaoTest.kt [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import org.hamcrest.CoreMatchers
+import org.hamcrest.MatcherAssert
+import org.junit.Assert
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class ProductsListDaoTest: DbTestMixin() {
+
+    private fun createCoursesItem(): List<CoursesItem>? {
```
[the remark this answers is not in the document]
Comment (bharathwaaj) [reply core/src/androidTest/java/in/testpress/core/database/ProductsListDaoTest.kt:17]: This is a fixture. Rename it accordingly.

File: core/src/androidTest/java/in/testpress/core/database/ProductsListDaoTest.kt [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+                descriptionHtml = "Course",currentPrice = "100",slug = "softskill",startDate = "Jan 4"
+        ))
+    }
+
+    private fun createProductList(): ProductsListEntity {
+        return ProductsListEntity(id = 1, courses = createCoursesItem(), products = createProductsItem())
+    }
+
+    @Test
+    fun readWrite() {
```
Comment (bharathwaaj) [inline core/src/androidTest/java/in/testpress/core/database/ProductsListDaoTest.kt:40]: This test case is not needed

File: core/src/main/java/in/testpress/util/Converters.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    fun stringToCoursesItem(value: String?): List<CoursesItem?>? {
+        val listType = object : TypeToken<ArrayList<CoursesItem?>?>() {}.type
+        return Gson().fromJson(value, listType)
+    }
+
+    @TypeConverter
+    @JvmStatic
+    fun coursesItemToString(courses: List<CoursesItem?>?): String? {
+        val gson = Gson()
+        return gson.toJson(courses)
```
Comment (bharathwaaj) [inline core/src/main/java/in/testpress/util/Converters.kt:22]:
```suggestion
        return Gson().toJson(courses)
```

File: core/src/main/java/in/testpress/util/Converters.kt [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    fun stringToProductsItem(value: String?): List<ProductsItem?>? {
+        val listType = object : TypeToken<ArrayList<ProductsItem?>?>() {}.type
+        return Gson().fromJson(value, listType)
+    }
+
+    @TypeConverter
+    @JvmStatic
+    fun productsItemToString(products: List<ProductsItem?>?): String? {
+        val gson = Gson()
+        return gson.toJson(products)
```
Comment (bharathwaaj) [inline core/src/main/java/in/testpress/util/Converters.kt:36]:
```suggestion
        return Gson().toJson(products)
```

**Review by bharathwaaj — CHANGES_REQUESTED** [review]

**Review by bharathwaaj — APPROVED** [review]

**Pull request merged**